cpp/src/copying/concatenate.cu

```diff
@@ -166,7 +166,7 @@ __global__ void fused_concatenate_kernel(column_device_view const* input_views,
   auto const output_size = output_view.size();
   auto* output_data      = output_view.data<T>();
 
-  size_type output_index     = threadIdx.x + blockIdx.x * blockDim.x;
+  std::size_t output_index   = threadIdx.x + blockIdx.x * blockDim.x;
```

Comment:
Using `output_index` in `size_t` may lead to some casting problem later, since it will be compared with many other indices in `int32_t`. We just need to detect if `output_index` here is overflow. So:
```suggestion
  std::size_t tmp_index       = threadIdx.x + blockIdx.x * blockDim.x;
  if(tmp_index > std::numeric_limit<size_type>::max()) { return; }
  
  std::size_t output_index   = static_cast<size_type>(tmp_index);
```

Reply:
Let's just keep with using a 64 bit integer type, but let's use `int64_t` instead of `size_t` to avoid unsigned comparison warnings.

Reply:
@ttnghia I moved to `int64_t` given @jrhemstad's comment. I looked at the comparisons with `output_index`, and I believe types are upgraded correctly so we shouldn't see an issue at least as the code is written.

Reply:
Yeah, agree :)

Reply:
Isn't `decltype(threadIdx.x + blockIdx.x * blockDim.x) == int32_t`, because the operands are that type? 
Would it be preferable to cast an operand to `int64_t`?
```c++
auto const tmp_index = threadIdx.x + static_cast<int64_t>(blockIdx.x)*blockDim.x;
```

Reply:
So the bug is actually later, and I don't think we can overflow the usual `threadIdx.x + blockIdx.x * blockDim.x` math given the 256 threads per block, the way we compute the number of blocks, and row limits in cuDF, if I understand your comment @mythrocks. 

The real bug is later on at:
```
output_index += blockDim.x * gridDim.x;
```
as mentioned in the PR description, that is because `output_index + (blockDim.x * gridDim.x)` can be larger than `int32_t`.

But this brings up a point. `fused_concatenate` checks [here](https://github.com/rapidsai/cudf/blob/2527ab1649e375999afff35747c113728307348b/cpp/src/copying/concatenate.cu#L225) that `output_size` needs to be less than size_type::max (`2^31-1`). That seems like a bug, but I am probably mistaken. Should it be `output_size - 1 <= static_cast<std::size_t>(std::numeric_limits<size_type>::max()`? I.e. if my type supports numbers from 0 to 10, and I have 11 rows, that should be OK (I can address with index 0..10). The reason I ask for this is because I wanted to test I can materialize the maximum number of rows from the concatenate.

Reply:
I think there's an implicit limitation to `INT_MAX - 1` elements in a column for several functions.

Reply:
> I think there's an implicit limitation to INT_MAX - 1 elements in a column for several functions.

Ok, so it is a cuDF limitation or bug? I would have expected the number of elements in the column to be up to `INT_MAX`, so that we can index using `0..INT_MAX - 1`.

Also I think the change should be to make the assert: 

``
output_size <= static_cast<std::size_t>(std::numeric_limits<size_type>::max()
```

But I am confused by your comment @jrhemstad. Should I make this change?

Reply:
> But I am confused by your comment @jrhemstad. Should I make this change?

Feel free to change it. I'm just saying, don't expect a column with `INT_MAX` elements to work everywhere :wink: